dm/relay/local_reader_test.go

```diff
@@ -1190,3 +1184,342 @@ func (t *testReaderSuite) createMetaFile(c *C, relayDirPath, binlogFileName stri
+	panic("should be used")
+}
+
+func (m *mockFileWriterForActiveTest) IsActive(uuid, filename string) (bool, int64) {
+	v := m.cases[m.cnt]
+	m.cnt++
+	return v.active, v.offset
+}
+
+func (t *testReaderSuite) TestwaitBinlogChanged(c *C) {
+	var (
+		relayFiles = []string{
+			"mysql-bin.000001",
+			"mysql-bin.000002",
+		}
+		binlogPos  = uint32(4)
+		binlogGTID = "ba8f633f-1f15-11eb-b1c7-0242ac110002:1"
+		relayPaths = make([]string, len(relayFiles))
+		data       = []byte("meaningless file content")
+		size       = int64(len(data))
+	)
+
+	// create relay log dir
+	subDir := c.MkDir()
+	// join the file path
+	for i, rf := range relayFiles {
+		relayPaths[i] = filepath.Join(subDir, rf)
+		f, _ := os.Create(relayPaths[i])
+		_ = f.Close()
+	}
+
+	rotateRelayFile := func(filename string) {
+		meta := LocalMeta{BinLogName: filename, BinLogPos: binlogPos, BinlogGTID: binlogGTID}
+		metaFile, err2 := os.Create(path.Join(subDir, utils.MetaFilename))
+		c.Assert(err2, IsNil)
+		err := toml.NewEncoder(metaFile).Encode(&meta)
+		c.Assert(err, IsNil)
+		_ = metaFile.Close()
+	}
+
+	// meta not found
+	{
+		cfg := &BinlogReaderConfig{RelayDir: "", Flavor: gmysql.MySQLFlavor}
+		r := newBinlogReaderForTest(log.L(), cfg, false, "")
+		t.setActiveRelayLog(r.relay, "next", "next", 0)
+		state := t.createBinlogFileParseState(c, subDir, relayFiles[0], 0, true)
+		needSwitch, reParse, err := r.waitBinlogChanged(context.Background(), state)
+		c.Assert(needSwitch, IsFalse)
+		c.Assert(reParse, IsFalse)
+		c.Assert(err, NotNil)
+		c.Assert(err, ErrorMatches, ".*no such file or directory*")
+	}
+
+	// write meta
+	rotateRelayFile(relayFiles[0])
+
+	// relay file not found
+	{
+		cfg := &BinlogReaderConfig{RelayDir: "", Flavor: gmysql.MySQLFlavor}
+		r := newBinlogReaderForTest(log.L(), cfg, false, "")
+		t.setActiveRelayLog(r.relay, "next", "next", 0)
+		state := &binlogFileParseState{
+			relayLogDir:  subDir,
+			relayLogFile: "not-exist-file",
+		}
+		needSwitch, reParse, err := r.waitBinlogChanged(context.Background(), state)
+		c.Assert(needSwitch, IsFalse)
+		c.Assert(reParse, IsFalse)
+		c.Assert(err, NotNil)
+		c.Assert(err, ErrorMatches, ".*no such file or directory*")
+	}
+
+	// create the first relay file
+	err1 := os.WriteFile(relayPaths[0], data, 0o600)
+	c.Assert(err1, IsNil)
+	// rotate relay file
+	rotateRelayFile(relayFiles[1])
+
+	// file decreased when meta changed
+	{
+		cfg := &BinlogReaderConfig{RelayDir: "", Flavor: gmysql.MySQLFlavor}
+		r := newBinlogReaderForTest(log.L(), cfg, false, "")
+		t.setActiveRelayLog(r.relay, "next", "next", 0)
+		state := t.createBinlogFileParseState(c, subDir, relayFiles[0], size+100, true)
+		needSwitch, reParse, err := r.waitBinlogChanged(context.Background(), state)
+		c.Assert(needSwitch, IsFalse)
+		c.Assert(reParse, IsFalse)
+		c.Assert(err, NotNil)
+		c.Assert(err, ErrorMatches, ".*file size of relay log.*become smaller.*")
```

Comment:
```suggestion
		c.Assert(terror.ErrRelayLogFileSizeSmaller.Equal(err), IsTrue)
```
and rest, this is better for review and debug